 ### 추가문제
##### Q. titanic_df의 각 칼럼별 null의 갯수를 for 반복문을 사용해서 구한 후 출력하세요.
###### (missing_count는 예외처리 하고 처리 방식은 0 을 출력함)

In [1]:
import seaborn as sns
import pandas as pd
import numpy as np

titanic_df = sns.load_dataset('titanic')
try:
    for column in titanic_df.columns :
        print(column,titanic_df[column].isnull().sum(axis=0))
except IndexError as e:
    print('인덱스에러',e) # 1번 방법

# 2번 방법
# for i in range(len(titanic_df)):
#     print(titanic_df.iloc[:,i].isnull().sum())

survived 0
pclass 0
sex 0
age 177
sibsp 0
parch 0
fare 0
embarked 2
class 0
who 0
adult_male 0
deck 688
embark_town 2
alive 0
alone 0


### 데이터 소스 : titanic_df = sns.load_dataset('titanic')
### Q. embark_town 열의 829행의 NaN 데이터의 앞, 뒤를 포함하는 3개행을 출력하세요.

In [2]:
import seaborn as sns
import pandas as pd
import numpy as np
titanic_df = sns.load_dataset('titanic')
display(titanic_df.head(3))
print(titanic_df.embark_town.iloc[828:831],'\n')

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True

828    Queenstown
829           NaN
830     Cherbourg
Name: embark_town, dtype: object 



### Q. embark_town 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 대체하세요.

In [3]:
nan_embark_town = titanic_df['embark_town'].value_counts(dropna=False)
display(nan_embark_town) # 가장 많이 출현한 값 : Southampton

titanic_df['embark_town'].replace(np.nan, 'Southampton', inplace=True)
titanic_df.dropna(subset=['embark_town'], axis=0, inplace=True)
titanic_df.embark_town.value_counts()

Southampton    644
Cherbourg      168
Queenstown      77
NaN              2
Name: embark_town, dtype: int64

Southampton    646
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64

### Q. titanic_df에서 'age' 열에 나이 데이터가 없는 모든 행을 삭제 후 df_age 이름으로 저장하고 행의 갯수를 구하세요.  데이터 소스 : titanic_df = sns.load_dataset('titanic')

In [4]:
titanic_df = sns.load_dataset('titanic')
display(titanic_df['age'].isnull().sum()) # 나이데이터에 NaN 수
df_age = titanic_df.dropna(how='any',subset=['age'])
df_age.shape # 행의 갯수 714개

177

(714, 15)

### Q. age 열의 NaN값을 다른 나이 데이터의 평균으로 변경한 후 'age' 열의 첫 10개 데이터만 정수로 출력하세요.

In [5]:
print('평균값:',titanic_df['age'].mean())
titanic_df['age'].replace(np.nan, titanic_df['age'].mean(), inplace=True)
titanic_df['age'].head(10).astype('int')

평균값: 29.69911764705882


0    22
1    38
2    26
3    35
4    35
5    29
6    54
7     2
8    27
9    14
Name: age, dtype: int32

### Q. 1 ~ 2사이의 정수로 (5,3) 2차원 배열을 랜덤하게 만든 후 DataFrame으로 불러와 df에 저장하고 df에서 중복 행을 제거한 후 df2에 저장 후 출력하세요.(columns = ['a','b','c'])

In [6]:
df = pd.DataFrame(np.random.randint(1,2,size=(5,3)), columns=['a','b','c'])
display(df)
df2 = df.drop_duplicates()
display(df2)

a  b  c
0  1  1  1
1  1  1  1
2  1  1  1
3  1  1  1
4  1  1  1

a  b  c
0  1  1  1

### Q. df에서 b, c열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.

In [7]:
df3=df.drop_duplicates(subset=['b','c'])
df3

a  b  c
0  1  1  1

### Q. titanic_df embark_town 열의 NaN값을 바로 앞에 있는 830행의 값으로 변경한 후 출력하세요.

In [8]:
titanic_df = sns.load_dataset('titanic')
display(titanic_df.embark_town.iloc[828:831])
titanic_df['embark_town'].fillna(method='bfill', inplace=True)
display(titanic_df.embark_town.iloc[828:831])

828    Queenstown
829           NaN
830     Cherbourg
Name: embark_town, dtype: object

828    Queenstown
829     Cherbourg
830     Cherbourg
Name: embark_town, dtype: object

In [9]:
# 데이터 프레임 생성
import pandas as pd
auto_df = pd.read_csv('auto-mpg.csv', header=None)
auto_df.columns = ['mpg','cylinders','displacement','horsepower','weight','acceleration','model year','origin','name'] 

auto_df.head(3)

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite

### Q. horsepower 열의 누락 데이터 '?'을 삭제한 후 NaN 값의 갯수를 출력하세요.

In [10]:
auto_df['horsepower'].replace('?', np.nan, inplace=True) # ?값을 NaN 으로 대체
auto_df.dropna(subset=['horsepower'], axis=0, inplace=True) # 누락된 데이터 삭제
auto_df['horsepower'].isnull().sum()

0

### Q. horsepower 열의 누락 데이터('?') 삭제하고 실수형으로 변환하세요.

In [11]:
auto_df['horsepower'] = auto_df['horsepower'].astype('float')
auto_df['horsepower'].head()

0    130.0
1    165.0
2    150.0
3    150.0
4    140.0
Name: horsepower, dtype: float64

### Q. Titanic 가공한 내역을 정리하고 이를 쉽게 재사용할 후 있도록 종합 전처리 함수를 만드세요.(def transform_features(df):)

In [23]:
# titanic_df = sns.load_dataset('titanic')
# 앞에서 설정한 Data Preprocessing 함수 호출

from sklearn.preprocessing import LabelEncoder
import numpy as np
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Null 처리 함수
def fillna(df):
    df['Age'].fillna(df['Age'].mean(), inplace=True)
    df['Cabin'].fillna('N', inplace=True)
    df['Embarked'].fillna('N', inplace=True)
    return df

# 머신러닝 알고리즘에 불필요한 속성 제거(PassengerId, Name, Ticket)
def drop_features(df):
    df.drop(['PassengerId','Name','Ticket'], axis=1, inplace=True)
    return df

# 레이블 인코딩
def format_features(df):
    df['Cabin'] = df['Cabin'].str[:1]
    features = ['Cabin','Sex','Embarked']
    for feature in features:
        le = LabelEncoder()
#         le = le.fit(df[feature])
        df[feature] = le.fit_transform(df[feature])
    return df

# 앞에서 설정한 Data Preprocessing 함수 호출
def transform_features(df):
    df = fillna(df)
    df = drop_features(df)
    df = format_features(df)
    return df

titanic_df = pd.read_csv('train_df.csv')
transform_features(titanic_df.head())

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Cabin  Embarked
0         0       3    1  22.0      1      0   7.2500      1         1
1         1       1    0  38.0      1      0  71.2833      0         0
2         1       3    0  26.0      0      0   7.9250      1         1
3         1       1    0  35.0      1      0  53.1000      0         1
4         0       3    1  35.0      0      0   8.0500      1         1